### PR TITLE
 Memory: Clean up locking around allocator implementations

### PR DESCRIPTION
**Pull request by evan-charmworks**
- Use of the ptmalloc3 "gnu" allocator has been made more modular so that it is now easier to add new malloc backends besides it and the OS allocator, as well as to allow further work toward using it as a memory pool for Isomalloc, HAPI, or more, if desired.
- Underlying malloc implementations are now exposed in two ways: `mm_*` symbols lock if the implementation requires it (such as ptmalloc3), while `mm_impl_*` symbols do no locking (unless the implementation, such as the OS, does its own). This allows our memory layers to handle locking as efficiently as possible.
- `-memory leak`, `-memory paranoid`, and `-memory verbose` are now thread-safe in all circumstances, including the default `-memory os-*`, by handling locking as part of their implementations. `-memory charmdebug` is also closer to being thread-safe, but it would require more changes outside of this PR's scope.
- Previously, `-memory gnu-*` always locked around the calls into our Converse memory layers, which allowed thread-safe usage of most memory modes, but also added unnecessary locking to ones that do not need it. This has been eliminated.
- The hook approach to intercepting malloc has been modularized so it does not require manual placement of `BEFORE`/`AFTER_MALLOC_CALL` at call sites. With this in place, in theory it may not require much work to add `libmemory-hooks-` variants of modes besides `charmdebug` to `src/scripts/Makefile`, though this is likely of interest to no one, especially since the hook method is not thread-safe.

### CODE REVIEW

**Comment by evan-charmworks** [comment]
This PR makes the following improvements:

- Use of the ptmalloc3 "gnu" allocator has been made more modular so that it is now easier to add new malloc backends besides it and the OS allocator, as well as to allow further work toward using it as a memory pool for Isomalloc, HAPI, or more, if desired.
- Underlying malloc implementations are now exposed in two ways: `mm_*` symbols lock if the implementation requires it (such as ptmalloc3), while `mm_impl_*` symbols do no locking (unless the implementation, such as the OS, does its own). This allows our memory layers to handle locking as efficiently as possible.
- `-memory leak`, `-memory paranoid`, and `-memory verbose` are now thread-safe in all circumstances, including the default `-memory os-*`, by handling locking as part of their implementations. `-memory charmdebug` is also closer to being thread-safe, but it would require more changes outside of this PR's scope.
- Previously, `-memory gnu-*` always locked around the calls into our Converse memory layers, which allowed thread-safe usage of most memory modes, but also added unnecessary locking to ones that do not need it. This has been eliminated.
- The hook approach to intercepting malloc has been modularized so it does not require manual placement of `BEFORE`/`AFTER_MALLOC_CALL` at call sites. With this in place, in theory it may not require much work to add `libmemory-hooks-` variants of modes besides `charmdebug` to `src/scripts/Makefile`, though this is likely of interest to no one, especially since the hook method is not thread-safe.

File: src/conv-core/memory-lock.C [hunk 1]
```diff
@@ -13,72 +7,72 @@ static void meta_init(char **argv)
 void *meta_malloc(size_t size)
 {
   void *result;
-  MEM_LOCK_AROUND( result = mm_malloc(size); )
+  MEM_LOCK_AROUND( result = mm_impl_malloc(size); )
```
Comment (minitu) [inline src/conv-core/memory-lock.C]: Does `MEM_LOCK_AROUND` still work with the macro definition removed?

Comment (evan-charmworks) [reply src/conv-core/memory-lock.C]: Yes, it is in memory.C now.

**Comment by stwhite91** [comment]
> This PR makes the following improvements:
> 
> * Use of the ptmalloc3 "gnu" allocator has been made more modular so that it is now easier to add new malloc backends besides it and the OS allocator, as well as to allow further work toward using it as a memory pool for Isomalloc, HAPI, or more, if desired.
> * Underlying malloc implementations are now exposed in two ways: `mm_*` symbols lock if the implementation requires it (such as ptmalloc3), while `mm_impl_*` symbols do no locking (unless the implementation, such as the OS, does its own). This allows our memory layers to handle locking as efficiently as possible.
> * `-memory leak`, `-memory paranoid`, and `-memory verbose` are now thread-safe in all circumstances, including the default `-memory os-*`, by handling locking as part of their implementations. `-memory charmdebug` is also closer to being thread-safe, but it would require more changes outside of this PR's scope.
> * Previously, `-memory gnu-*` always locked around the calls into our Converse memory layers, which allowed thread-safe usage of most memory modes, but also added unnecessary locking to ones that do not need it. This has been eliminated.
> * The hook approach to intercepting malloc has been modularized so it does not require manual placement of `BEFORE`/`AFTER_MALLOC_CALL` at call sites. With this in place, in theory it may not require much work to add `libmemory-hooks-` variants of modes besides `charmdebug` to `src/scripts/Makefile`, though this is likely of interest to no one, especially since the hook method is not thread-safe.

Could you put this in the commit message body?

**Comment by evan-charmworks** [comment]
> Could you put this in the commit message body?

Done

**Comment by minitu** [comment]
Approved. So how can I use this in HAPI with `cudaMalloc` as a backend?

**Comment by evan-charmworks** [comment]
> Approved. So how can I use this in HAPI with `cudaMalloc` as a backend?

I'm still looking at that. This commit is cleanup that allows further work toward this goal.

**Comment by evan-charmworks** [comment]
This diff may be easier to view by running `git fetch && git diff -w --color-moved origin/master...origin/evan/cleanup-memlock`.

**Comment by evan-charmworks** [comment]
Could this get another review when someone has the chance? I would like to push other PRs that build on the work done here.

**Pull request merged**